src/mz/src/command/profile.rs

```diff
@@ -0,0 +1,295 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License in the LICENSE file at the
+// root of this repository, or online at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Implementation of the `mz profile` command.
+//!
+//! Consult the user-facing documentation for details.
+
+use std::{io::Write, str::FromStr};
+
+use mz_frontegg_auth::AppPassword;
+use mz_frontegg_client::client::app_password::CreateAppPasswordRequest;
+use mz_frontegg_client::client::{Client as AdminClient, Credentials};
+use mz_frontegg_client::config::{
+    ClientBuilder as AdminClientBuilder, ClientConfig as AdminClientConfig,
+};
+
+use mz_cloud_api::config::DEFAULT_ENDPOINT;
+use serde::{Deserialize, Serialize};
+use tabled::Tabled;
+use tokio::{select, sync::mpsc};
+use url::Url;
+
+use crate::ui::OptionalStr;
+use crate::{
+    config_file::TomlProfile,
+    context::{Context, ProfileContext},
+    error::Error,
+    server::server,
+};
+
+/// Opens the default web browser in the host machine
+/// and awaits a single request containing the profile's app password.
+pub async fn init_with_browser() -> Result<AppPassword, Error> {
+    // Bind a web server to a local port to receive the app password.
+    let (tx, mut rx) = mpsc::unbounded_channel();
+    let (server, port) = server(tx);
+
+    // Build the login URL
+    let mut url = DEFAULT_ENDPOINT.clone();
```

Comment:
This function should take the hidden `cloud-endpoint` flag as a parameter to allow using mz against staging.

Reply:
This works
```patch
diff --git a/src/mz/src/bin/mz/command/profile.rs b/src/mz/src/bin/mz/command/profile.rs
index 9fcff1289..b5b00a281 100644
--- a/src/mz/src/bin/mz/command/profile.rs
+++ b/src/mz/src/bin/mz/command/profile.rs
@@ -89,6 +89,7 @@ pub async fn run(mut cx: Context, cmd: ProfileCommand) -> Result<(), Error> {
                 &mut cx,
                 profile,
                 *no_browser,
+                endpoint.cloud_endpoint.clone(),
                 endpoint.admin_endpoint.clone(),
             )
             .await
diff --git a/src/mz/src/command/profile.rs b/src/mz/src/command/profile.rs
index 30cb716f2..50a181f8e 100644
--- a/src/mz/src/command/profile.rs
+++ b/src/mz/src/command/profile.rs
@@ -42,13 +42,13 @@ use crate::{
 
 /// Opens the default web browser in the host machine
 /// and awaits a single request containing the profile's app password.
-pub async fn init_with_browser() -> Result<AppPassword, Error> {
+pub async fn init_with_browser(cloud_endpoint: Option<Url>) -> Result<AppPassword, Error> {
     // Bind a web server to a local port to receive the app password.
     let (tx, mut rx) = mpsc::unbounded_channel();
     let (server, port) = server(tx);
 
     // Build the login URL
-    let mut url = DEFAULT_ENDPOINT.clone();
+    let mut url = cloud_endpoint.unwrap_or_else(|| DEFAULT_ENDPOINT.clone());
     url.path_segments_mut()
         .expect("constructor validated URL can be a base")
         .extend(&["account", "login"]);
@@ -136,11 +136,12 @@ pub async fn init(
     scx: &mut Context,
     profile_name: Option<String>,
     no_browser: bool,
+    cloud_endpoint: Option<Url>,
     admin_endpoint: Option<Url>,
 ) -> Result<(), Error> {
     let app_password = match no_browser {
         true => init_without_browser(admin_endpoint.clone()).await?,
-        false => init_with_browser().await?,
+        false => init_with_browser(cloud_endpoint.clone()).await?,
     };
 
     let new_profile = TomlProfile {
@@ -148,7 +149,7 @@ pub async fn init(
         vault: None,
         region: None,
         admin_endpoint: admin_endpoint.map(|url| url.to_string()),
-        cloud_endpoint: None,
+        cloud_endpoint: cloud_endpoint.map(|url| url.to_string()),
     };
 
     scx.config_file()
diff --git a/src/mz/src/command/region.rs b/src/mz/src/command/region.rs
index d53679d48..76d19f56c 100644
--- a/src/mz/src/command/region.rs
+++ b/src/mz/src/command/region.rs
@@ -24,7 +24,7 @@ use crate::{context::RegionContext, error::Error};
 use mz_cloud_api::client::{cloud_provider::CloudProvider, environment::Environment};
 use serde::{Deserialize, Serialize};
 use tabled::Tabled;
-use tokio::time::{sleep};
+use tokio::time::sleep;
 
 /// Enable a region in the profile organization.
 ///
@@ -52,14 +52,14 @@ pub async fn enable(cx: RegionContext) -> Result<(), Error> {
         match cx.get_environment(region.clone()).await {
             Ok(environment) => {
                 if environment.resolvable {
-                    break Ok(environment)
+                    break Ok(environment);
                 }
-            },
+            }
             Err(e) => {
                 if tries == 10 {
                     break Err(e);
                 }
-            },
+            }
         }
         sleep(Duration::from_secs(10)).await;
     }?;
```